server/channels/app/post.go

```diff
@@ -797,27 +792,36 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
 	}
 
 	permalinkPreviewedPost := post.GetPreviewPost()
-	for _, cm := range channelMembers {
-		if permalinkPreviewedPost != nil {
-			post.Metadata.Embeds[0].Data = permalinkPreviewedPost
-		}
-
-		postForUser := a.sanitizePostMetadataForUserAndChannel(c, post, permalinkPreviewedPost, permalinkPreviewedChannel, cm.UserId)
+	if permalinkPreviewedPost != nil {
+		post.Metadata.Embeds[0].Data = permalinkPreviewedPost
+	} else {
+		return false, nil
+	}
 
-		// Using DeepCopy here to avoid a race condition
-		// between publishing the event and setting the "post" data value below.
-		messageCopy := message.DeepCopy()
-		broadcastCopy := messageCopy.GetBroadcast()
-		broadcastCopy.UserId = cm.UserId
-		messageCopy.SetBroadcast(broadcastCopy)
+	postCopy := post.Clone()
 
-		postJSON, jsonErr := postForUser.ToJSON()
-		if jsonErr != nil {
-			mlog.Warn("Failed to encode post to JSON", mlog.Err(jsonErr))
+	if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, post.UserId, permalinkPreviewedChannel) {
+		post.Metadata.Embeds[0].Data = nil
+	}
+	// Using DeepCopy here to avoid a race condition
+	// between publishing the event and setting the "post" data value below.
+	messageCopy := message.DeepCopy()
+	broadcastCopy := messageCopy.GetBroadcast()
+	broadcastCopy.ChannelHook = func(userID string, ev *model.WebSocketEvent) bool {
+		if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, userID, permalinkPreviewedChannel) {
+			postCopy.Metadata.Embeds[0].Data = nil
```

Comment:
Is `postCopy` the same instance each time we call this hook? At first glance, it seems like the first user without permission to the channel will modify the shared `postCopy` for all subsequent users whether or not they have access.

Reply:
In my testing, that didn't happen. However, I think I might have been just lucky. I will verify this deeply. It looks like I need to clone inside the hook.